alacritty_terminal/src/locale.rs

```diff
@@ -59,21 +59,32 @@ pub fn set_locale_environment() {
     let locale_c_str = CString::new(locale_id.to_owned()).unwrap();
     let locale_ptr = locale_c_str.as_ptr();
     let locale_id = unsafe {
-        // save a copy of original setting
-        let original = setlocale(LC_CTYPE, null());
-        let saved_original = if original.is_null() {
-            CString::new("").unwrap()
-        } else {
-            CStr::from_ptr(original).to_owned()
-        };
         // try setting `locale_id`
         let modified = setlocale(LC_CTYPE, locale_ptr);
         let result = if modified.is_null() { String::new() } else { locale_id };
-        // restore original setting
-        setlocale(LC_CTYPE, saved_original.as_ptr());
         result
     };
 
+    // Provide a default value for LC_CTYPE if needed
+    unsafe {
+        let orginal = setlocale(LC_CTYPE, null());
+        let ctype = if orginal.is_null() {
+            CString::new("C").unwrap()
+        } else {
+            CStr::from_ptr(orginal).to_owned()
+        };
+
+        let ctype_string = ctype.into_string().unwrap();
+        match ctype_string.as_str() {
+            "C" => {
+                let utf8 = CString::new("UTF-8").unwrap();
+                setlocale(LC_CTYPE, utf8.as_ptr());
+                env::set_var("LC_CTYPE", utf8.into_string().unwrap());
+            },
+            _ => (),
+        }
```

Comment:
```suggestion
        let ctype = setlocale(LC_CTYPE, null());
        if !original.is_null() && CStr::from_ptr(ctype).to_string_lossy() == "C" {
            let locale = "UTF-8";
            setlocale(LC_CTYPE, CString::new(locale).unwrap());
            env::set_var("LC_CTYPE", locale);
        }
```

I think something like this should work a bit better.

However I question how well this will work by just retrieving the locale. I can test this on macOS tomorrow, but does this deal with setting the locale from the CLI correctly?

Reply:
Hm, @chrisduerr. It seems like passing `LC_CTYPE` like `LC_CTYPE=en_US.UTF8 alacritty` does not pass it correctly. Any idea why?

Also your suggested code has some errors; i fixed those – what about the following?
```rust
let ctype = setlocale(LC_CTYPE, null());
if !ctype.is_null() && CStr::from_ptr(ctype).to_string_lossy() == "C" {
    let locale = "UTF-8";
    setlocale(LC_CTYPE, CString::new(locale).unwrap().as_ptr());
    env::set_var("LC_CTYPE", locale);
}

```

Reply:
> Hm, @chrisduerr. It seems like passing LC_CTYPE like LC_CTYPE=en_US.UTF8 alacritty does not pass it correctly. Any idea why?

I think you might be confusing locale with the environment variable. The environment variable is processed with `env::var("LC_CTYPE")`, while `setlocale(LC_CTYPE, null())` just reads the currently used locale.

> Also your suggested code has some errors; i fixed those – what about the following?

Yeah, should be fine. Was mostly about the general idea than the character for character implementation. The functionality is the bigger problem.